### PR TITLE
[RLlib; docs] Redo `rllib-algorithms.rst` page.

### PR DESCRIPTION
**Pull request by sven1977**
Redo `rllib-algorithms.rst` page.

* New algo architecture diagrams.
* New algo overview table (updated and more suited to new API stack).
* Do-over of algo descriptions.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/rllib/images/algos/ppo-architecture.svg
Comment (simonsays1980) [inline doc/source/rllib/images/algos/ppo-architecture.svg:1]: The `gym.vector` might led users to believe that this holds for SARL and MARL while the latter does not support vector envs, yet.

File: rllib/algorithms/ppo/ppo.py [hunk 1]
```diff
@@ -70,37 +70,39 @@ class PPOConfig(AlgorithmConfig):
...
         config = PPOConfig()
-        config = config.training(gamma=0.9, lr=0.01, kl_coeff=0.3,
-            train_batch_size=128)
-        config = config.resources(num_gpus=0)
-        config = config.env_runners(num_env_runners=1)
+        config.environment("CartPole-v1")
+        config.training(
+            gamma=0.9, lr=0.01, kl_coeff=0.3, train_batch_size_per_learner=256
+        )
+        config.resources(num_gpus=0)
```
Comment (simonsays1980) [inline rllib/algorithms/ppo/ppo.py:79]: This is still ambiguous. We have `num_gpus` in `resources`, `num_gpus_per_learner` in `learners` and `num_gpus_per_env_runner`. What is this `gpu` for? Is it a gpu for the driver? Is it the total number of gpus available to env runners and learners?

File: doc/source/rllib/rllib-algorithms.rst [hunk 2]
```diff
@@ -7,236 +7,176 @@
...
-   :members: training
+Overview
+~~~~~~~~
+
++-----------------------------------------------------------------------------+------------------------------+------------------------------------+--------------------------------+
+| **Algorithm**                                                               | **Single- and Multi-agent**  | **Multi-GPU (multi-node)**         | **Action Spaces**              |
++-----------------------------------------------------------------------------+------------------------------+------------------------------------+--------------------------------+
+| **On-Policy**                                                                                                                                                                    |
++-----------------------------------------------------------------------------+------------------------------+------------------------------------+--------------------------------+
+| :ref:`PPO (Proximal Policy Optimization) <ppo>`                             | |single_agent| |multi_agent| | |multi_gpu| |multi_node_multi_gpu| | |cont_actions| |discr_actions| |
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-algorithms.rst:18]: Make more clear, what is the difference between multi-gpu and multi-node-multi-gpu.

File: doc/source/rllib/rllib-algorithms.rst [hunk 3]
```diff
@@ -7,236 +7,176 @@
...
++-----------------------------------------------------------------------------+------------------------------+------------------------------------+--------------------------------+
+| **Off-Policy**                                                                                                                                                                   |
++-----------------------------------------------------------------------------+------------------------------+------------------------------------+--------------------------------+
+| :ref:`DQN/Rainbow (Deep Q Networks) <dqn>`                                  | |single_agent| |multi_agent| | |multi_gpu| |multi_node_multi_gpu| |                |discr_actions| |
++-----------------------------------------------------------------------------+------------------------------+------------------------------------+--------------------------------+
+| :ref:`SAC (Soft Actor Critic) <sac>`                                        | |single_agent| |multi_agent| | |multi_gpu| |multi_node_multi_gpu| | |cont_actions|                 |
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-algorithms.rst:24]: That's actually not correct. Mulit-learner settings are not available for SAC  - due to the multiple optimizers per learner.

File: doc/source/rllib/rllib-algorithms.rst [hunk 4]
```diff
@@ -7,236 +7,176 @@
...
 
-In offline RL, the algorithm has no access to an environment, but can only sample from a fixed dataset of pre-collected state-action-reward tuples.
-In particular, CQL (Conservative Q-Learning) is an offline RL algorithm that mitigates the overestimation of Q-values outside the dataset distribution via
-conservative critic estimates. It does so by adding a simple Q regularizer loss to the standard Bellman update loss.
-This ensures that the critic does not output overly-optimistic Q-values. This conservative
-correction term can be added on top of any off-policy Q-learning algorithm (here, we provide this for SAC).
+    **PPO architecture:** In a training iteration, PPO performs three major steps: sampling a set of episodes or episode fragments (1),
+    converting these into a train batch and updating the model(s) using a clipped objective and multiple SGD passes over this batch (2),
+    and synching the weights from the Learners back to the EnvRunners (3).
+    PPO scales out on both axes, supporting multiple EnvRunners for sample collection and multiple GPU- or CPU-based Learner
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-algorithms.rst:64]: Maybe `Learner-S`?

File: doc/source/rllib/rllib-algorithms.rst [hunk 5]
```diff
@@ -7,236 +7,176 @@
...
-We include an asynchronous variant of Proximal Policy Optimization (PPO) based on the IMPALA architecture. This is similar to IMPALA but using a surrogate policy loss with clipping. Compared to synchronous PPO, APPO is more efficient in wall-clock time due to its use of asynchronous sampling. Using a clipped loss also allows for multiple SGD passes, and therefore the potential for better sample efficiency compared to IMPALA. V-trace can also be enabled to correct for off-policy samples.
+**Tuned examples:**
+`PongDeterministic-v4 <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/dqn/pong-dqn.yaml>`__,
+`Rainbow configuration <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/dqn/pong-rainbow.yaml>`__,
+`{BeamRider,Breakout,Qbert,SpaceInvaders}NoFrameskip-v4 <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/dqn/atari-dqn.yaml>`__,
+`with Dueling and Double-Q <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/dqn/atari-duel-ddqn.yaml>`__,
+`with Distributional DQN <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/dqn/atari-dist-dqn.yaml>`__.
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-algorithms.rst:108]: Maybe we want to mention the rainbow architecture earlier above.

File: doc/source/rllib/rllib-algorithms.rst [hunk 6]
```diff
@@ -7,236 +7,176 @@
...
-Qbert          11085           14247           ~14000
-SpaceInvaders  671             944             ~800
-=============  ==============  ==============  ==================
+RLlib's soft-actor critic implementation is ported from the `official SAC repo <https://github.com/rail-berkeley/softlearning>`__ to better integrate with RLlib APIs.
+Note that SAC has two fields to configure for custom models: ``policy_model_config`` and ``q_model_config``, the ``model`` field of the config is ignored.
 
+Tuned examples (continuous actions):
+`Pendulum-v1 <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/sac/pendulum-sac.yaml>`__,
+`HalfCheetah-v3 <https://github.com/ray-project/ray/blob/master/rllib/tuned_examples/sac/halfcheetah-sac.yaml>`__,
+Tuned examples (discrete actions):
```
Comment (simonsays1980) [inline doc/source/rllib/rllib-algorithms.rst:145]: Discrete actions are not implemented in the new stack.

**Review by simonsays1980 — APPROVED** [review]
LGTM. SOme nits here and there. The descriptions are actually really good. Short, precise, straight-to-the-point.

File: doc/source/rllib/images/algos/ppo-architecture.svg
Comment (sven1977) [reply doc/source/rllib/images/algos/ppo-architecture.svg:1]: Ah, good catch. I think, we'll have to fix this limitation on MAEnvRunner soon :|

File: doc/source/rllib/rllib-algorithms.rst (hunk 4 above)
Comment (sven1977) [reply doc/source/rllib/rllib-algorithms.rst:64]: fixed

File: doc/source/rllib/rllib-algorithms.rst (hunk 3 above)
Comment (sven1977) [reply doc/source/rllib/rllib-algorithms.rst:24]: fixed

File: doc/source/rllib/rllib-algorithms.rst (hunk 2 above)
Comment (sven1977) [reply doc/source/rllib/rllib-algorithms.rst:18]: Good point!

File: doc/source/rllib/rllib-algorithms.rst (hunk 6 above)
Comment (sven1977) [reply doc/source/rllib/rllib-algorithms.rst]: True! Good catch.

Comment (sven1977) [reply doc/source/rllib/rllib-algorithms.rst]: The table has this correctly.

Comment (simonsays1980) [reply doc/source/rllib/rllib-algorithms.rst]: Ah alright. Must have overseen it there.

File: doc/source/rllib/rllib-algorithms.rst (hunk 2 above)
Comment (sven1977) [reply doc/source/rllib/rllib-algorithms.rst:22]: Done.

**Pull request merged**